### PR TITLE
Traffic pcap

### PR DESCRIPTION
**Pull request by outscale-toa**

### CODE REVIEW

File: api/client/nic.cc [hunk 1]
```diff
@@ -389,6 +389,7 @@ static int SubNicSg(int argc, char **argv, const GlobalOptions &options) {
 }
 
 NicAddOptions::NicAddOptions() {
+    packet_trace = "false";
```
Comment (outscale-mgo) [inline api/client/nic.cc]: packet_trace = "";

File: api/protocol/message_0.proto [hunk 2]
```diff
@@ -184,6 +184,8 @@ message MessageV0 {
     optional Type type = 9 [default = VHOST_USER_SERVER];
     // Nic path may be set depending of NIC's type
     optional string path = 10;
+    // dump VM traffic in pcap file
+    optional bool packet_trace = 11 [default = false];
```
Comment (outscale-mgo) [inline api/protocol/message_0.proto]: s/[default = false]//

File: api/protocol/message_0.proto [hunk 3]
```diff
@@ -278,6 +280,8 @@ message MessageV0 {
     // Replace Security Groups of a NIC
     // To remove all security groups, set a empty string
     repeated string security_group = 4;
+    // Update dumping vm traffic in pcap file
+    optional bool packet_trace = 5 [default = false];
```
Comment (outscale-mgo) [inline api/protocol/message_0.proto]: s/[default = false]//

File: api/server/api_0.cc [hunk 4]
```diff
@@ -777,6 +779,10 @@ bool Api0::Convert(const MessageV0_Nic &nic_message, app::Nic *nic_model) {
     nic_model->ip_anti_spoof = false;
     if (nic_message.has_ip_anti_spoof())
         nic_model->ip_anti_spoof = nic_message.ip_anti_spoof();
+    // Packet trace
+    nic_model->packet_trace = false;
```
Comment (outscale-mgo) [inline api/server/api_0.cc]: `nic_model->packet_trace = app::config.packet_trace`

File: api/server/api_0.cc [hunk 5]
```diff
@@ -806,6 +812,13 @@ bool Api0::Convert(const MessageV0_NicUpdateReq &nic_update_message,
     } else {
         nic_update_model->has_ip_anti_spoof = false;
     }
+    // packet trace
+    if (nic_update_message.has_packet_trace()) {
+        nic_update_model->packet_trace = nic_update_message.packet_trace();
+        nic_update_model->has_packet_trace = true;
+    } else {
+        nic_update_model->has_packet_trace = false;
```
Comment (outscale-mgo) [inline api/server/api_0.cc]: `nic_update_model->packet_trace = app::config.packet_trace`

File: api/server/graph.cc [hunk 6]
```diff
@@ -732,6 +738,73 @@ void Graph::NicConfigAntiSpoof(const app::Nic &nic, bool enable) {
     }
 }
 
+void Graph::EnablePacketTrace(const app::Nic &nic) {
+    Graph::GraphNic *g_nic = FindNic(nic);
+    std::string name;
+    bool enable = true;
+
```
Comment (outscale-mgo) [inline api/server/graph.cc]: check if tracing is alerady enable here please

File: api/server/graph.cc [hunk 7]
```diff
@@ -732,6 +738,73 @@ void Graph::NicConfigAntiSpoof(const app::Nic &nic, bool enable) {
...
+        g_nic->head = g_nic->sniffer;
+        LinkAndStalk(g_nic->antispoof, g_nic->vhost, g_nic->sniffer,
+                    enable);
+    }
+}
+
+void Graph::DisablePacketTrace(const app::Nic &nic) {
+    Graph::GraphNic *g_nic = FindNic(nic);
+    bool disable = false;
+
```
Comment (outscale-mgo) [inline api/server/graph.cc]: check if tracing is already disable here please

File: api/server/api.cc [hunk 8]
```diff
@@ -214,6 +214,13 @@ bool Api::ActionNicUpdate(const Api::NicUpdate &update,
         app::graph.NicConfigAntiSpoof(n, n.ip_anti_spoof);
     }
 
+    // Update packet trace if needed
+    if (update.has_packet_trace &&
```
Comment (outscale-mgo) [inline api/server/api.cc]: `if (update.has_packet_trace)` is enouth

File: api/client/nic.cc [hunk 9]
```diff
@@ -417,6 +418,9 @@ int NicAddOptions::Parse(int argc, char **argv) {
             vni = string(argv[i + 1]);
         else if (string(argv[i]) == "--bypass-filtering")
             bypass_filtering = "true";
+        else if ((string(argv[i]) == "--packet-trace") ||
+                 (string(argv[i]) == "-t"))
+            packet_trace = "packet_trace: " + std::string(argv[i + 1]);
```
Comment (outscale-mgo) [inline api/client/nic.cc]: you don't check what's inside argv[i + 1]
example: 
Here I can use `--packet_trace toto` and client will say nothing

File: api/server/api_0.cc [hunk 10]
```diff
@@ -806,6 +812,14 @@ bool Api0::Convert(const MessageV0_NicUpdateReq &nic_update_message,
     } else {
         nic_update_model->has_ip_anti_spoof = false;
     }
+    // packet trace
+    if (nic_update_message.has_packet_trace()) {
+        nic_update_model->packet_trace = nic_update_message.packet_trace();
+        nic_update_model->has_packet_trace = true;
+    } else {
+        nic_update_model->packet_trace = app::config.packet_trace;
```
Comment (outscale-mgo) [inline api/server/api_0.cc]: I think you can remove `nic_update_model->packet_trace = app::config.packet_trace`

File: api/server/graph.cc [hunk 11]
```diff
@@ -70,19 +70,12 @@ Graph::~Graph(void) {
...
-                         Graph::BrickShrPtr sniffer) {
-    if (app::config.packet_trace) {
-        if (pg_brick_chained_links(&app::pg_error, westBrick.get(),
-                                   sniffer.get(), eastBrick.get()) < 0) {
-            PG_ERROR_(app::pg_error);
-            return false;
-        }
+                         Graph::BrickShrPtr sniffer, bool enable) {
+    if (enable) {
+        link(westBrick, sniffer);
```
Comment (outscale-mgo) [inline api/server/graph.cc]: sadly, I don't think that's okay to use link

You need to use link, when updating tracing,  but not when crating a nic.
The code work, but as Graph::LinkAndStalk is used inside nic add, by doing that, you slow down pooling more that it should at every nic add.

So either you give link function pointer in parameter, so you can configure the way linkAndStalk will link, either you need to create a new linkAndStaklk.

File: api/server/graph.cc [hunk 12]
```diff
@@ -732,6 +731,75 @@ void Graph::NicConfigAntiSpoof(const app::Nic &nic, bool enable) {
     }
 }
 
+void Graph::EnablePacketTrace(const app::Nic &nic) {
+    Graph::GraphNic *g_nic = FindNic(nic);
+    std::string name;
+    if (!nic.packet_trace) {
+        app::log.Error("packet_trace has to be true!");
```
Comment (outscale-mgo) [inline api/server/graph.cc]: just return
 or Use log info

File: api/server/graph.cc [hunk 13]
```diff
@@ -732,6 +731,75 @@ void Graph::NicConfigAntiSpoof(const app::Nic &nic, bool enable) {
...
+        g_nic->head = g_nic->sniffer;
+        LinkAndStalk(g_nic->antispoof, g_nic->vhost, g_nic->sniffer,
+                    nic.packet_trace);
+    }
+}
+
+void Graph::DisablePacketTrace(const app::Nic &nic) {
+    Graph::GraphNic *g_nic = FindNic(nic);
+    if (nic.packet_trace) {
+        app::log.Error("packet_trace has to be false!");
```
Comment (outscale-mgo) [inline api/server/graph.cc]: same here

File: api/server/api.cc [hunk 14]
```diff
@@ -214,6 +214,12 @@ bool Api::ActionNicUpdate(const Api::NicUpdate &update,
         app::graph.NicConfigAntiSpoof(n, n.ip_anti_spoof);
     }
 
+    // Update packet trace if needed
+    if (update.has_packet_trace && update.packet_trace != n.packet_trace) {
+        n.packet_trace = update.packet_trace;
```
Comment (outscale-mgo) [inline api/server/api.cc]: well, you know it
`update.packet_trace != n.packet_trace` check is useless

File: api/client/nic.cc [hunk 15]
```diff
@@ -437,6 +441,8 @@ static void SubNicAddHelp(void) {
             << endl <<
         "    --enable-antispoof  enable antispoof protection (default: off)"
             << endl <<
+        "    -t true/false or --packet-trace true/false  enable traffic in" <<
```
Comment (outscale-mgo) [inline api/client/nic.cc]: replace "enable traffic in pcap file" by "choose whatever we trace a nic (default: use default server behaviour)"

File: tests/api/MessageV0/test_nic_00_out [hunk 16]
```diff
@@ -1,5 +1,5 @@
 messages {
-  revision: PROTO_REVISION
```
Comment (outscale-mgo) [inline tests/api/MessageV0/test_nic_00_out]: This is wrong
You need to update api/protocol/version

File: api/protocol/message_0.proto [hunk 17]
```diff
@@ -184,6 +184,8 @@ message MessageV0 {
     optional Type type = 9 [default = VHOST_USER_SERVER];
     // Nic path may be set depending of NIC's type
     optional string path = 10;
+    // dump VM traffic in pcap file
+    optional bool packet_trace = 11;
```
Comment (outscale-jju) [inline api/protocol/message_0.proto]: would't be interesting to set a string instead of a bool so the user can choose in which file he wants his pcap to be written ? If the user sets an empty string (but the field is set), then an automatic file could be chosen.

File: api/server/graph.cc [hunk 18]
```diff
@@ -732,6 +738,74 @@ void Graph::NicConfigAntiSpoof(const app::Nic &nic, bool enable) {
...
+        g_nic->head = g_nic->sniffer;
+        link(g_nic->antispoof, g_nic->sniffer);
+        link(g_nic->sniffer, g_nic->vhost);
+    }
+}
+
+void Graph::DisablePacketTrace(const app::Nic &nic) {
+    Graph::GraphNic *g_nic = FindNic(nic);
+    if (!nic.packet_trace) {
+        app::log.Debug("packet_trace has to be true on the nic to update!");
```
Comment (outscale-jju) [inline api/server/graph.cc]: I don't find the message clear, would say as a WARNING:
> packet trace option on nic XXX is already disabled
Same for enabling
Also, we can avoid using "!"

File: api/server/graph.h [hunk 19]
```diff
@@ -102,6 +115,7 @@ class Graph {
         VHOST_STOP,
         LINK,
         UNLINK,
+        UNLINK_EDGE,
```
Comment (outscale-jju) [inline api/server/graph.h]: Adding a new async action could be splitted in a separate commit I guess

File: tests/api/MessageV0/test_nic_00_out [hunk 20]
```diff
@@ -94,6 +96,7 @@ messages {
         bypass_filtering: false
         type: VHOST_USER_SERVER
         path: "/tmp/qemu-vhost-nic-2"
+        packet_trace: true
```
Comment (outscale-jju) [inline tests/api/MessageV0/test_nic_00_out]: please, put protocol update, API server update, client update and test in separate commits

File: api/protocol/message_0.proto (hunk 17 above)
Comment (outscale-toa) [reply api/protocol/message_0.proto]: I think we have to add another option for the path, because we trace all nics by default, and if i set this option to sting, we will not get option to disable packet tracing

File: api/server/graph.cc [hunk 21]
```diff
@@ -521,8 +521,10 @@ bool Graph::NicAdd(app::Nic *nic_) {
     }
     if (nic.packet_trace) {
         name = "sniffer-" + gn.id;
-        gn.pcap_file = fopen(("/tmp/butterfly-" + std::to_string(getpid()) +
-                              "-" + gn.id + ".pcap").c_str(), "w");
+        gn.packet_trace_path = nic.packet_trace_path;
+        gn.pcap_file = fopen((nic.packet_trace_path + "/butterfly-" +
```
Comment (outscale-mgo) [inline api/server/graph.cc]: so, titi.pcap will create "titi.pcap/butterfly-PID-ID"
sorry that's just not what was asked
the goal of nic.packet_trace_path is to allow the user to chose the full path of his pcap
not the directory where the pcap file will be found

Comment (outscale-toa) [reply api/server/graph.cc]: - ```titi.pcap will create "titi.pcap/butterfly-PID-ID"``` 
    Yes if "titi.pcap" is a directory. I think the PATH mean the directory where to find something, rigth?

- You said : "the goal of nic.packet_trace_path is to allow the user to chose the full path of his pcap not the directory where the pcap file will be found" 
   So you mean we have to change the PATH and rename the pcap files if i understand
